crates/rpc/rpc-eth-api/src/helpers/legacy_xlayer.rs

```diff
@@ -37,16 +28,23 @@ pub fn should_route_to_legacy(
 
 /// Check if a BlockId should be routed to legacy RPC based on cutoff_block
 #[inline]
-pub fn should_route_block_id_to_legacy(
+pub fn should_route_block_id_to_legacy<Provider>(
     legacy_client: Option<&std::sync::Arc<reth_rpc_eth_types::LegacyRpcClient>>,
-    block_id: Option<BlockId>,
-) -> bool {
-    if let Some(client) = legacy_client {
-        if let Some(BlockId::Number(number)) = block_id {
-            return should_route_to_legacy(Some(client), number);
-        }
-    }
-    false
+    provider: &Provider,
+    block_id: Option<&BlockId>,
+) -> Result<bool, ErrorObjectOwned>
+where
+    Provider: reth_storage_api::BlockNumReader,
+{
+    let Some(client) = legacy_client else {
+        return Ok(false);
+    };
```

Comment:
it is not recommanded to use `let else` here, because you destructuring the `Option` and then use `Some` to wrap it latter

```Rust
if legacy_client.is_none() {
    return Ok(false);
}
```

is cleaner